### PR TITLE
Hot Fix: `--config` is being ignored

### LINKED ISSUE #2277: [regression][compiler] `--config` doesn't work in `0.47.0`
## Problem

In `0.47.0`, specifying a `--config` through CLI doesn't work, and it will parse the `main.tsp` as config YAML which will result in errors like below:
```
C:/Users/xxxxxxxxxxx/autorest.csharp/node_modules/@azure-tools/cadl-ranch-specs/http/type/model/inheritance/enum-discriminator/main.tsp:1:2 - error file-load: end of the stream or a document separator is expected (15:9)

 12 | @doc("extensible enum type for discrimi ...
 13 | enum DogKind {
 14 |   @doc("Species golden")
 15 |   Golden: "golden",
--------------^
 16 | }
 17 |
> 1 | import "@typespec/http";
```

The problem is in the https://github.com/microsoft/typespec/blob/42d00cef2bd13ec70407ea77d5a0249cdaa3f154/packages/compiler/src/config/config-to-options.ts#L49-L52
Due to operator precedence，`options.configPath ?? entrypointStat.isDirectory()` is calculated first. So you'll always get `options.entrypoint` if `--config` is specified.
That can be confirmed by the transpiled codes:
> const configPath = ((_b = options.configPath) !== null && _b !== void 0 ? _b : entrypointStat.isDirectory())
        ? options.entrypoint
        : getDirectoryPath(options.entrypoint);

### PR DESCRIPTION
**Pull request by archerzz**
- fix an expression typo when determining the value of `configPath`
- add test cases

fix #2277

### CODE REVIEW

**Comment by timotheeguerin** [comment]
Should we do a hot fix for this?

File: packages/compiler/test/config/config.test.ts [hunk 1]
```diff
@@ -1,19 +1,25 @@
-import { deepStrictEqual, strictEqual } from "assert";
-import { dirname, join, resolve } from "path";
+import { deepStrictEqual, fail, strictEqual } from "assert";
+import { dirname, join, normalize, resolve } from "path";
```
Comment (timotheeguerin) [inline packages/compiler/test/config/config.test.ts:2]: Should try not to use anything from `path` anywhere but scripts. This cause issues with windows vs unix path and everything in the compiler should already have been normalized to unix paths. The compiler provide some path functions that can be used instead.

File: packages/compiler/test/config/config.test.ts [hunk 2]
```diff
@@ -148,3 +154,58 @@ describe("compiler: config file loading", () => {
...
+      const [options, diagnostics] = await resolveCompilerOptions(NodeHost, {
+        entrypoint: fullPath, // not really used here
+        configPath: fullPath,
+      });
+      return { options, diagnostics };
+    };
+
+    it("resolves full path to custom config file", async () => {
+      const config = await resolveOptions("custom/myConfig.yaml");
+      config.options.config = normalize(config.options.config!);
```
Comment (timotheeguerin) [inline packages/compiler/test/config/config.test.ts:172]: Like having to do that just so you can compare on both platform

File: packages/compiler/test/config/config.test.ts [hunk 3]
```diff
@@ -148,3 +154,58 @@ describe("compiler: config file loading", () => {
...
+    it("throws error when config file not found", async () => {
+      try {
+        await resolveOptions("not-found.yaml");
+      } catch (error) {
+        return;
+      }
+      fail("should throw error");
+    });
+
+    it("resolves yaml config file", async () => {
```
Comment (timotheeguerin) [inline packages/compiler/test/config/config.test.ts:195]: A bit confused by what this test is testing, shouldn't `--config` always only point to a config file? Do we allow it to be a directory too?

File: common/changes/@typespec/compiler/fix-config-option_2023-08-10-08-29.json [hunk 4]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@typespec/compiler",
+      "comment": "fix(config): --config is not adopted",
```
Comment (timotheeguerin) [inline common/changes/@typespec/compiler/fix-config-option_2023-08-10-08-29.json:5]: This shows up in the changelog, it should be a bit more descriptive too

**Comment by timotheeguerin** [comment]
Thanks for the Fix @archerzz I converted this to be an hotfix or I guess you will be prevented from migrating to 0.47. Also fixed up the comments I made on the PR and filed an issue about a behavior you added a test for which I feel is unexpected https://github.com/microsoft/typespec/issues/2282(Please comment on it if you are depending on it)

File: packages/compiler/test/config/config.test.ts (hunk 3 above)
Comment (timotheeguerin) [reply packages/compiler/test/config/config.test.ts]: Filed an issue to discuss this behavior https://github.com/microsoft/typespec/issues/2282. Please comment on it if it was something you expected to work and depended on.

**Review by timotheeguerin — APPROVED** [review]

**Comment by archerzz** [comment]
> Thanks for the Fix @archerzz I converted this to be an hotfix or I guess you will be prevented from migrating to 0.47. Also fixed up the comments I made on the PR and filed an issue about a behavior you added a test for which I feel is unexpected #2282 comment on it if you are depending on it)

Thanks!

**Pull request merged**